src/components/navigation/right-nav.tsx

```diff
@@ -55,36 +98,76 @@ export class RightNavComponent extends BaseComponent<IProps, {}> {
 
   private renderTabContents() {
     const {activeRightNavTab} = this.stores.ui;
-    const tabComponents: { [tab: string]: any } = {
-      "My Work": MyWorkComponent,
-      "Class Work": ClassWorkComponent,
-      "Class Logs": ClassLogsComponent
-    };
-    const _TabComponent = tabComponents[activeRightNavTab];
-    if (_TabComponent) {
-      return (
-        <div className="contents">
-          <_TabComponent scale={kRightNavItemScale} />
-        </div>
-      );
-    }
+    return (
+      <div className="contents">
+        { this.state.tabLoadAllowed[kMyWorkTab]
+          ? <div className={"container " + (activeRightNavTab === kMyWorkTab ? "enabled" : "disabled")}>
+              <MyWorkComponent scale={kRightNavItemScale}/>
+            </div>
+          : this.renderLoadingText(kMyWorkTab)
+        }
+        { this.state.tabLoadAllowed[kClassWorkTab]
+          ? <div className={"container " + (activeRightNavTab === kClassWorkTab ? "enabled" : "disabled")}>
+              <ClassWorkComponent scale={kRightNavItemScale}/>
+          </div>
+          : this.renderLoadingText(kClassWorkTab)
+        }
+        { this.state.tabLoadAllowed[kClassLogsTab]
+          ? <div className={"container " + (activeRightNavTab === kClassLogsTab ? "enabled" : "disabled")}>
+              <ClassLogsComponent scale={kRightNavItemScale}/>
+            </div>
+          : this.renderLoadingText(kClassLogsTab)
+        }
+      </div>
+    );
+  }
+
+  private renderLoadingText(tab: string) {
+    const {activeRightNavTab} = this.stores.ui;
+    return (
+      <div>
+        { tab === activeRightNavTab
+          ? <div className="loading">loading...</div>
+          : null
+        }
+      </div>
+    );
   }
 
   private handleTabClick = (tab: string) => {
     const { ui } = this.stores;
+    const navDoneExpanding = ui.rightNavExpanded;
     return (e: React.MouseEvent<HTMLDivElement>) => {
+      if (!navDoneExpanding) {
+        this.setState({navExpanding: true});
+      }
       if (ui.activeRightNavTab !== tab) {
         ui.setActiveRightNavTab(tab);
         this.stores.ui.toggleRightNav(true);
-      }
-      else {
+      } else {
         this.stores.ui.toggleRightNav();
       }
+      if (navDoneExpanding) {
+        this.updateComponentLoadAllowedState();
+      }
     };
   }
 
   private getTabId(tab: string) {
     return `rightNavTab${tab}`;
   }
 
+  private updateComponentLoadAllowedState = () => {
+    const { ui } = this.stores;
+    const tabLoadAllowed = this.state.tabLoadAllowed;
+    if (ui.activeRightNavTab === kMyWorkTab) {
+      tabLoadAllowed[kMyWorkTab] = true;
+    } else if (ui.activeRightNavTab === kClassWorkTab) {
+      tabLoadAllowed[kClassWorkTab] = true;
+    } else if (ui.activeRightNavTab === kClassLogsTab) {
+      tabLoadAllowed[kClassLogsTab] = true;
+    }
```

Comment:
Final nit: with the `tabLoadAllowed` flags, it seems like this could be reduced to something like:
```
const tabLoadAllowed = this.state.tabLoadAllowed;
tabLoadAllowed[ui.activeRightNavTab] = true;
this.setState({ tabLoadAllowed });

Reply:
👍